modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java

```diff
@@ -166,503 +152,196 @@ public int pages() {
 
-                        fileExists = true;
+    /**
+     * Gets or creates a new delta file, a new delta file will be created when the previous one is {@link #completeNewDeltaFile()
+     * completed}.
+     *
+     * <p>Thread safe.
+     *
+     * @param pageIndexesSupplier Page indexes supplier for the new delta file page store.
+     * @return Future that will be completed when the new delta file page store is created.
+     */
+    public CompletableFuture<DeltaFilePageStoreIo> getOrCreateNewDeltaFile(Supplier<int[]> pageIndexesSupplier) {
+        assert deltaFilePageStoreIoFactory != null;
 
-                        checkHeader(fileIo);
+        CompletableFuture<DeltaFilePageStoreIo> future = this.newDeltaFilePageStoreIoFuture;
 
-                        this.fileIo = fileIo;
+        if (future != null) {
+            return future;
+        }
 
-                        if (interrupted) {
-                            Thread.currentThread().interrupt();
-                        }
+        if (!NEW_DELTA_FILE_PAGE_STORE_IO_FUTURE.compareAndSet(this, null, future = new CompletableFuture<>())) {
+            // Another thread started creating a delta file.
+            return newDeltaFilePageStoreIoFuture;
+        }
 
-                        break;
-                    } catch (ClosedByInterruptException e) {
-                        interrupted = true;
+        int nextIndex = deltaFilePageStoreIos.isEmpty() ? 0 : deltaFilePageStoreIos.get(0).fileIndex() + 1;
 
-                        Thread.interrupted();
-                    }
-                }
-            } catch (IOException e) {
-                try {
-                    if (fileIo != null) {
-                        fileIo.close();
-                    }
-                } catch (IOException e0) {
-                    e.addSuppressed(e0);
-                }
+        DeltaFilePageStoreIo deltaFilePageStoreIo = deltaFilePageStoreIoFactory.create(nextIndex, pageIndexesSupplier.get());
 
-                throw e;
-            }
-        } finally {
-            readWriteLock.writeLock().unlock();
-        }
-    }
+        // Should add to the head, since read operations should always start from the most recent.
+        deltaFilePageStoreIos.add(0, deltaFilePageStoreIo);
 
-    private static int calcCrc32(ByteBuffer pageBuf, int pageSize) {
-        try {
-            pageBuf.position(0);
+        future.complete(deltaFilePageStoreIo);
 
-            return FastCrc.calcCrc(pageBuf, pageSize);
-        } finally {
-            pageBuf.position(0);
-        }
+        return future;
     }
 
     /**
-     * Returns file page store path.
+     * Completes the {@link #getOrCreateNewDeltaFile(Supplier) creation} of a new delta file.
+     *
+     * <p>Thread safe.
+     *
+     * @throws IgniteInternalCheckedException If failed.
      */
-    public Path filePath() {
-        return filePath;
-    }
+    public void completeNewDeltaFile() throws IgniteInternalCheckedException {
```

Comment:
Future is already completed in `getOrCreateNewDeltaFile`, isn't it? I don't understand what's happening here. We clearly need more comments or a more clear code structure

Reply:
I will add more comments later.